python/cudf/bindings/csv.pyx

```diff
@@ -0,0 +1,423 @@
+        that starts before or at the end of the range, even if it ends after
+        the end of the range.
+    skip_blank_lines : bool, default True
+        If True, discard and do not parse empty lines
+        If False, interpret empty lines as NaN values
+    comment : char, default None
+        Character used as a comments indicator. If found at the beginning of a
+        line, the line will be ignored altogether.
+    na_values : list, default None
+        Values to consider as invalid
+    keep_default_na : bool, default True
+        Whether or not to include the default NA values when parsing the data.
+    na_filter : bool, default True
+        Detect missing values (empty strings and the values in na_values).
+        Passing False can improve performance.
+    prefix : str, default None
+        Prefix to add to column numbers when parsing without a header row
+    index_col : int or string, default None
+        Column to use as the row labels
+
+    Returns
+    -------
+    GPU ``DataFrame`` object.
+
+    Examples
+    --------
+
+    Create a test csv file
+
+    >>> import cudf
+    >>> filename = 'foo.csv'
+    >>> lines = [
+    ...   "num1,datetime,text",
+    ...   "123,2018-11-13T12:00:00,abc",
+    ...   "456,2018-11-14T12:35:01,def",
+    ...   "789,2018-11-15T18:02:59,ghi"
+    ... ]
+    >>> with open(filename, 'w') as fp:
+    ...     fp.write('\\n'.join(lines)+'\\n')
+
+    Read the file with ``cudf.read_csv``
+
+    >>> cudf.read_csv(filename)
+      num1                datetime text
+    0  123 2018-11-13T12:00:00.000 5451
+    1  456 2018-11-14T12:35:01.000 5784
+    2  789 2018-11-15T18:02:59.000 6117
+    """
+
+    if delim_whitespace:
+        if delimiter is not None:
+            raise ValueError("cannot set both delimiter and delim_whitespace")
+        if sep != ',':
+            raise ValueError("cannot set both sep and delim_whitespace")
+
+    # Alias sep -> delimiter.
+    if delimiter is None:
+        delimiter = sep
+
+    if dtype is not None:
+        if isinstance(dtype, collections.abc.Mapping):
+            dtype_dict = True
+        elif isinstance(dtype, collections.abc.Iterable):
+            dtype_dict = False
+        else:
+            msg = '''dtype must be 'list like' or 'dict' '''
+            raise TypeError(msg)
+        if names is not None and len(dtype) != len(names):
+            msg = '''All column dtypes must be specified.'''
+            raise TypeError(msg)
+
+    nvtx_range_push("CUDF_READ_CSV", "purple")
+
+    cdef csv_read_arg csv_reader = csv_read_arg()
+
+    # Populate csv_reader struct
+    if is_file_like(filepath_or_buffer):
+        if compression == 'infer':
+            compression = None
+        buffer = filepath_or_buffer.read()
+        # check if StringIO is used
+        if hasattr(buffer, 'encode'):
+            buffer_as_bytes = buffer.encode()
+        else:
+            buffer_as_bytes = buffer
+        buffer_data_holder = <char*>buffer_as_bytes
```

Comment:
Do I need to free this later? The `buffer_as_bytes` Python object will get garbage collected.

Reply:
This returns the pointer held by the `bytes` object under the hood. So it is already accounted for by Python. IOW no there is no need to free this pointer.

Reply:
Thanks @jakirkham!